ocaml/typing/ctype.ml

```diff
@@ -2115,36 +2116,72 @@ let rec estimate_type_jkind env ty =
 
        This, however, still allows sort variables to get instantiated. *)
     Jkind jkind
-  | Tvar { jkind } -> TyVar (jkind, ty)
+  | Tvar { jkind } -> TyVar (ty, jkind)
   | Tarrow _ -> Jkind (Primitive.value ~why:Arrow)
   | Ttuple _ -> Jkind (Primitive.value ~why:Tuple)
+  | Tunboxed_tuple ltys -> Product (List.map snd ltys)
   | Tobject _ -> Jkind (Primitive.value ~why:Object)
   | Tfield _ -> Jkind (Primitive.value ~why:Tfield)
   | Tnil -> Jkind (Primitive.value ~why:Tnil)
   | (Tlink _ | Tsubst _) -> assert false
   | Tunivar { jkind } -> Jkind jkind
-  | Tpoly (ty, _) -> estimate_type_jkind env ty
+  | Tpoly (ty, _) -> estimate_type_jkind_head env ty
   | Tpackage _ -> Jkind (Primitive.value ~why:First_class_module)
 
+(* We parameterize [estimate_type_jkind] and [jkind_of_result] by a function
+   [expand_components] because some callers want expansion of types and others
+   don't. *)
+let rec estimate_type_jkind env ~expand_components ty =
+  jkind_of_result env ~expand_components (estimate_type_jkind_head env ty)
+
+and jkind_of_result env ~expand_components jkind_head =
+  match jkind_head with
+  | Jkind l -> l
+  | TyVar (_, l) -> l
+  | Product tys ->
+    Jkind.Primitive.product ~why:Unboxed_tuple
+      (List.map (fun ty ->
+         estimate_type_jkind env ~expand_components (expand_components ty)
+       ) tys)
+
 (**** checking jkind relationships ****)
 
+(* This type represents jkind constraints that we weren't able prove yet, but
+   which might succeed either by further expanding the type or by modifying a
+   type variable (in which case the [ty] will be a type variable). *)
+type type_jkind_sub_possible_failure =
+  { estimate : Jkind.t;
+    bound : Jkind.t;
+    ty : type_expr }
+
 type type_jkind_sub_result =
   | Success
-    (* The [Type_var] case might still be "success"; caller should check.
-       We don't just report success here because if the caller unifies the
-       tyvar, error messages improve. *)
-  | Type_var of Jkind.t * type_expr
-  | Missing_cmi of Jkind.t * Path.t
-  | Failure of Jkind.t
+  | Type_vars of type_jkind_sub_possible_failure list
+  (* [Type_vars] indicates a possible success - the caller should check, and
+     possibly update the variables *)
+  | Missing_cmi of Path.t
+  | Failure of type_jkind_sub_possible_failure
+  (* [Failure] indicates we couldn't verify the constraint.  However, further
+     expansion the type might make it possible to verify. *)
 
 let type_jkind_sub env ty jkind =
```

Comment:
After reading this all, I'm left with the feeling that there is probably a cleaner algorithm underneath this all. I see a few bits that don't quite smell right:
* The refactoring you suggest in the CR in `process_product_results`.
* The fact that the jkind estimate is returned and usually stored in the result.
* The fact that there is a failure-with-the-possibility of improvement, and then there is a hard failure -- yet this algorithm doesn't differentiate them. For example, if `is_nary_product` encounters a `float64`, no amount of expansion will help. It might be nice if a discovery like this just stopped the expansion.

None of these are terribly bad. But my overall sense here is that time spent here will hopefully yield code that is simpler and closer to "obviously no bugs" than "no obvious bugs".

Reply:
I confess I spent a fair amount of time on this and thought I had ended up with an algorithm that was pretty simple.  These three functions each do, in my mind, a clear and necessary thing - it would help to know what you think a simpler algorithm would look like.  Though I agree some of the comments could be better!

A few thoughts:

> The fact that the jkind estimate is returned and usually stored in the result.

I think these are two different things.  These functions return the jkind estimate for the entire type they are passed.  They also return a result.  That result may indicate we failed to prove several different constraints for several different types, and contains the estimates for those types (which are pieces of the original type, but not the same as the original type).  Of course, at the bottom of the recursion where we have non-composite layouts, the failure information and the estimate must be about the same type.

I did consider whether the estimate is needed at all in `type_jkind_sub_possible_failure`.  In practice, it's only really needed in the `Type_vars` case, but having one type to represent all potentially recoverable failures seems nice.

> The fact that there is a failure-with-the-possibility of improvement, and then there is a hard failure -- yet this algorithm doesn't differentiate them. For example, if is_nary_product encounters a float64, no amount of expansion will help. It might be nice if a discovery like this just stopped the expansion.

This is true!  To be clear, the old algorithm also suffered from this - checking whether something with layout `float64` had layout `value` would always result in fully expanding the type.  Our thought was in error cases like these, performance isn't so important - I am tempted to say that bad perf in the failure case is a good trade for a nice clear algorithm.  But I suppose it is the case that some kind checking failures are not actually errors (like mode crossing checks, and probably some gadt stuff), so I will make a ticket to explore this in the future.

Reply:
Yes, good points. I don't have a clear idea of what exactly I want here.... just a feeling that there is something yet simpler inside trying to get out. But we don't have to find that simpler thing (if it even exists) now -- I'd be happy to merge this as-is, especially if this is already the result of whittling. And that's true about the failure case being problematic previously.